org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js

```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
     }
 }
 
+/**
+ * If the user provides a unit, validate the value is within the units range.
+ *  Range is defined by [min/max][Heating/Cooling]SetPoint if available, otherwise:
+ *    c [9-32]
+ *    f [50-90]
+ * 
+ * If the unit is provided and within range, returns provided unit 
+ *  otherwise an exception is thrown.
+ * 
+ * If no unit is provided and the value is within one of the two ranges,
+ *  the unit of the valid range will be returned, otherwise an 
+ *  exception is thrown.
+ *  * 
+ * @param {*} resourceSchema 
+ * @param {*} providerSchema 
+ */
+function getValidatedUnit(resourceSchema, attributes) {
+    var value = resourceSchema.temperature;
+    if (resourceSchema.hasOwnProperty('units') && resourceSchema.units != null) {
+        var unit = resourceSchema.units.toLowerCase();
+        var providerUnit = getUnitSafe(attributes);
+        var min = getMinHeatingSetpoint(attributes);
+        var max = getMaxCoolingSetpoint(attributes);
+        if (unit === 'c' && providerUnit === 'f') {
+            value = celsiusToFahrenheit(value);
+        } else if (unit === 'f' && providerUnit === 'c') {
+            value = fahrenheitToCelsius(value);
+        }
+        if (value >= min && value <= max) {
+            return unit;
+        }
+        throw new OpenT2TError(440, "Invalid temperature (" + value + ") for unit (" + unit + "[" + min + ", " + max + "])");
+    } else {
+        var providerUnit = getUnitSafe(attributes);
+        var min = getMinHeatingSetpoint(attributes);
+        var max = getMaxCoolingSetpoint(attributes);
+        if (value >= min && value <= max) {
+            return providerUnit;
+        }
+        if (providerUnit === 'c') {
+            min = celsiusToFahrenheit(min);
+            max = celsiusToFahrenheit(max);
+            if (value >= min && value <= max) {
+                return 'f';
+            }
+        } else if (providerUnit === 'f') {
+            min = fahrenheitToCelsius(min);
+            max = fahrenheitToCelsius(max);
+            if (value >= min && value <= max) {
+                return 'c';
+            }
+        }
+        throw new OpenT2TError(440, "Temperature outside supported range (" + value + ")");
+    }
+}
+
+/**
+ * Given a target temperature, compute a target 
+ * low and target high centerd around it.
+ * @param {*} resourceSchema 
+ * @param {*} stateReader 
+ */
+function getTargetTemperatureRange(resourceSchema, attributes) {
+    var result = {};
```

Comment:
do this at line 355 i.e. closer to where it's needed

Reply:
Fixed.